### PR TITLE
Make aria2 secret a random value

### LINKED ISSUE #425: SECURITY: Public secret for Aria2c RPC command module
Simply put, the secret for the aria2 command is in the code, publicly visible (Unless I am mistaken in my C++ reading, and that default is never used).  That means any program capable of sending requests to localhost can order the Aria2 server to do whatever it wants.  That (AFAIK) includes javascript executed by websites, meaning a user with kiwix installed can be made to download arbitrary files just by visiting a website.

The RPC module of Aria2 is complex, and I don't have the time at the moment to examine it in great detail.  While it doesn't permit you to directly run arbitrary commands as download hooks, it is highly complicated and could have an arbitrary code execution issue.  Even if there isn't one, an attacker still has access to a powerful tool, and could probably force all kinds of nasty behaviors.

The fix to this is fairly easy: preprocess the secret in.  If the server is shut down on every update, simply set it to a compile-time random constant,  which would make the attack far harder.  I am not a security person, so I'm no expert on keeping these keys secret, but I do think that it should be at least more than just a string literal.

### PR DESCRIPTION
**Pull request by juuz0**
Fixes #425

### CODE REVIEW

**Comment by kelson42** [comment]
@juuz0 I guess this needs a PR on kiwix/kiwix-desktop side as well and that this libkiwix PR is insufficient to adapt kiwix-desktop.

**Comment by juuz0** [comment]
@kelson42 kiwix-desktop works fine without any changes though.. Not sure if this affects kiwix-desktop in any way.

**Comment by kelson42** [comment]
@juuz0 I'm sorry, I didn't had a look in detail but I though we talk about https://aria2.github.io/manual/en/html/aria2c.html?highlight=secret#cmdoption-rpc-secret... so it shoukd have an impact on kiwix-desktop side!?

**Comment by juuz0** [comment]
@kelson42 No, I don't think so. In Kiwix desktop we use the Downloader class (which uses Aria2 class and methods) so I don't think there should be any changes needed there

File: src/aria2.h [hunk 1]
```diff
@@ -37,6 +37,7 @@ class Aria2
 
     std::string addUri(const std::vector<std::string>& uri, const std::vector<std::pair<std::string, std::string>>& options = {});
     std::string tellStatus(const std::string& gid, const std::vector<std::string>& statusKey);
+    std::string makeAriaSecretToken();
```
Comment (kelson42) [inline src/aria2.h:40]: Better call something linke `getNewRpcSecret()`

File: src/aria2.cpp [hunk 2]
```diff
@@ -195,6 +197,14 @@ std::string Aria2::tellStatus(const std::string& gid, const std::vector<std::str
   return doRequest(methodCall);
 }
 
+std::string Aria2::makeAriaSecretToken()
+{
+  unsigned seed = std::chrono::system_clock::now().time_since_epoch().count();
+  std::string letters("0123456789abcdefghijklmnopqrstuvqxyzABCDEFGHIJKLMNOPQRSTUVWXYZ");
```
Comment (kelson42) [inline src/aria2.cpp:203]: For me this does not work, somehow the aria2c operation get stuck in Kiwix-Desktop. After testing around, it seems that the string is too long! If I put `0123456789abcdefghijklmnopqrstuvqxyzA`. It works fine.

**Review by kelson42 — CHANGES_REQUESTED** [review]
@juuz0 See my two comments, but if these details are fixed, then we should be good.

File: src/aria2.h [hunk 3]
```diff
@@ -37,6 +37,7 @@ class Aria2
 
     std::string addUri(const std::vector<std::string>& uri, const std::vector<std::pair<std::string, std::string>>& options = {});
     std::string tellStatus(const std::string& gid, const std::vector<std::string>& statusKey);
+    std::string getNewRpcSecret();
```
Comment (veloman-yunkan) [inline src/aria2.h:40]: Declare this as a static member function

**Review by kelson42 — APPROVED** [review]
LGTM

File: src/aria2.cpp [hunk 4]
```diff
@@ -195,6 +197,14 @@ std::string Aria2::tellStatus(const std::string& gid, const std::vector<std::str
   return doRequest(methodCall);
 }
 
+std::string Aria2::getNewRpcSecret()
+{
+  unsigned seed = std::chrono::system_clock::now().time_since_epoch().count();
+  std::string letters("0123456789abcdefghijklmnopqrstuvqxyzABCDEFGHIJKLMNOPQRSTUVWXYZ");
+  std::shuffle(letters.begin(), letters.end(), std::default_random_engine(seed));
+  return letters.substr(0, 9);
```
Comment (veloman-yunkan) [inline src/aria2.cpp:205]: This is not much better than simply converting `seed` to a string. The output is fully determined by the `seed` value (unless the underlying random engine also adds some host-dependent salt).

I think that returning the stringified `seed` value provides an adequate level of security. The extra code only creates a false impression of higher security.

Comment (kelson42) [reply src/aria2.cpp:205]: @veloman-yunkan  I don't understand your comment. `seed` is the clock, so it is not random and can not be used as random value. The shuffled string is not directly depending on `seed` but on the pseudo random generator of `std::default_random_engine` (itself seeded by `seed`). So, "yes" this is pseudo-random, but this is the usual way of doing to get a random value.

Comment (veloman-yunkan) [reply src/aria2.cpp:205]: @kelson42 As long as you know the value of `seed` you can find out the result of `std::shuffle()`, can't you?

Comment (kelson42) [reply src/aria2.cpp:205]: @veloman-yunkan Now I understand your comment. To answer your question, you might know or not. I don't know the details of implementation of the pseudo-random generator. But I still prefer this way as just putting the unix date (even if I understand your point). Maybe you have an idea to easily improve the quality of the seed?

Comment (veloman-yunkan) [reply src/aria2.cpp:205]: > But I still prefer this way as just putting the unix date

That's my concern. The result of the `shuffle()` is no better than the timestamp used as a seed value. If you are uncomfortable with the unix date returned directly you can reverse or square it (but that is intended only to illustrate my point).

> Maybe you have an idea to easily improve the quality of the seed?

I don't believe that this is the case where we should go into such lengths. Security measures should be proportional to the risks. I think that we have a lot of similar vulnerabilities in other places.

Comment (veloman-yunkan) [reply src/aria2.cpp:205]: BTW, why not taking advantage of the [`gen_uuid()`](https://github.com/kiwix/libkiwix/blob/b70c92cadefdb10295ad8bb34c992407519d193a/src/tools/otherTools.cpp#L368) function from `src/tools/otherTools.cpp`?

Comment (juuz0) [reply src/aria2.cpp:205]: Does `gen_uuid()` not suffer from the same problem of creating a false impression?
If we can get the random string, we can also get its uuid?

Comment (veloman-yunkan) [reply src/aria2.cpp:205]: If called with an empty string, `gen_uuid()` generates a random UUID based on the wall time and process CPU time values. Even though it is not an ideal source of random strings, the advantage of reusing it is that we don't create a new way of producing random strings. If we ever decide that our random string generation must be improved we will have to do it in a single place. 

Note that you may need to drop the dashes from the result (if they are not allowed in the token strings accepted by aria2).

File: src/aria2.cpp [hunk 5]
```diff
@@ -3,12 +3,14 @@
 #include "aria2.h"
 #include "xmlrpc.h"
 #include <iostream>
+#include <algorithm>
```
Comment (veloman-yunkan) [inline src/aria2.cpp:6]: This header is not needed either

**Review by veloman-yunkan — CHANGES_REQUESTED** [review]
This is almost an approval. Please rebase-fixup for the next iteration to get the formal approval.

File: src/aria2.cpp (hunk 5 above)
Comment (juuz0) [reply src/aria2.cpp:6]: It is for std::remove

Comment (veloman-yunkan) [reply src/aria2.cpp:6]: Indeed.

**Review by veloman-yunkan — APPROVED** [review]

File: src/aria2.h (hunk 3 above)
Comment (kelson42) [reply src/aria2.h]: done

**Comment by kelson42** [comment]
Rebasing and merging...

**Pull request merged**